src/Dialect/ONNX/ONNXOpsHelper.cpp

```diff
@@ -0,0 +1,42 @@
+//====------ ONNXOpsHelper.cpp - Helper functions for ONNX dialects -------===//
+//
+// Copyright 2019 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file contains helper functions for lowering ONNX ops to Krnl Dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Dialect/ONNX/ONNXOpsHelper.hpp"
+
+// Identity affine
+using namespace mlir;
+AffineMap getIdentityDimMap(Builder &builder) {
+  return AffineMap::get(1, 0, {builder.getAffineDimExpr(0)});
+}
+
+// Pool/conv affine
+// dim =
+//   let numerator = (input + pad - (kernel - 1) * dilation - 1)
+//   in let denominator = stride
+//      in
+//        if (ceilMode)
+//          ceil(numerator / denominator) + 1
+//        else
+//          floor(numerator / denominator) + 1
+AffineMap getConvDimMap(Builder &builder, bool ceilMode) {
+  AffineExpr input = builder.getAffineDimExpr(0);
+  AffineExpr kernel = builder.getAffineSymbolExpr(0);
```

Comment:
Maybe I'm missing something, why do we need to make `kernel` an affine symbol expr? The fact that the value of `kernel` can be obtained from an operation attribute implies that the value of `kernel` is always a compile-time constant. Which makes me feel like this should always be a constant, which means we should simply use:

```
int64_t kernel = {value obtained from reading into an attribute};
```

And use this constant value to construct the affine mapping.

Reply:
Right, we can simplify the constant part of the formula into a single constant. and use it to construct the affine mapping. Here, I explicitly put `kernel, `pad`, `stride`, `dilation` to the affine map in order to make the formula appear in the program so that it is easy to check the correctness. For example, in `lit` tests, we can check how we compute an output dimension by looking at the map instead of a simplified constant.

Reply:
I'm afraid that when we try to tile this loop, a boundary condition as complicated as this one will give the polyhedral analysis tool a very hard time to figure out the tiled boundaries, which are often one level of complexity beyond those before tiling.

Regardless, I'm open to merging this patch as is, and come back later for modification if we actually run into troubles. Let's also ask @AlexandreEichenberger for more comments.